core/src/services/s3/lister.rs

```diff
@@ -100,20 +100,21 @@ impl oio::PageList for S3Lister {
         };
         ctx.token = output.next_continuation_token.clone().unwrap_or_default();
 
+        let mut prefix_existing = false;
         for prefix in output.common_prefixes {
             let de = oio::Entry::new(
                 &build_rel_path(&self.core.root, &prefix.prefix),
                 Metadata::new(EntryMode::DIR),
             );
 
             ctx.entries.push_back(de);
+            prefix_existing = true;
         }
 
         for object in output.contents {
             let path = build_rel_path(&self.core.root, &object.key);
 
-            // s3 could return the dir itself in contents.
-            if path == self.path || path.is_empty() {
+            if path.is_empty() {
```

Comment:
there is one corner case here, what should `list("")` should return ? `""` or `"/"`. `""` seems strange, but `"/"` is not consistent with others results, since other paths don't have a leading slash

Reply:
> `""` seems strange, but `"/"` is not consistent with others results, since other paths don't have a leading slash

`/` is somehow special in opendal which represets `root` folder of current operator. And `list("")` is the same with `list("/")`. So if `/` does exist, I think it's fine to return it too.